### PR TITLE
[DebugInfo] Add flag to enable function-level debug line attribution

### PR DESCRIPTION
**Pull request by alx32**
This patch introduces a new flag, `-mllvm -emit-func-debug-line-table-offsets`, which enables the generation of additional debug information to facilitate the attribution of debug line data to their corresponding functions (`DW_TAG_subprogram`).

**Context:**
Currently, Clang generates function line information in the `debug_line` section without directly linking the line entries to their originating functions. This becomes problematic when post-compile tools, such as the Identical Code Folding (ICF) in the LLD linker, merge multiple functions together, resulting in overlapping address ranges and conflicting line information.

To address this issue, this patch adds a new attribute, `DW_AT_META_stmt_sequence`, to each `DW_TAG_subprogram`, which represents an offset into the line table where the line data for the subprogram begins. Additionally, a `DW_LNE_end_sequence` is added to the line table to mark the end of the line information for a particular subprogram. These changes enable the correct attribution of line entries to their originating functions, even when functions share the same address space after optimization.

**NOTE:** While this change is functional, it requires additional tooling support to be fully utilized, particularly for symbol resolution in cases where multiple functions share the same address space. This will be addressed in future changes.

RFC: https://discourse.llvm.org/t/rfc-new-dwarf-attribute-for-symbolication-of-merged-functions/79434

**Previous changes:** [#93137](https://github.com/llvm/llvm-project/pull/93137) - Added an LLD option to generate debug information for ICF'ed functions.

### CODE REVIEW

File: llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll [hunk 1]
```diff
@@ -0,0 +1,68 @@
+; RUN: llc -mtriple=i686-w64-mingw32 -o %t -filetype=obj %s
+; RUN: llvm-dwarfdump -v -all %t | FileCheck %s -check-prefix=NO_STMT_SEQ
+
+; RUN: llc -mtriple=i686-w64-mingw32 -o %t -filetype=obj %s -emit-func-debug-line-table-offsets
+; RUN: llvm-dwarfdump -v -all %t | FileCheck %s -check-prefix=STMT_SEQ
+
+; NO_STMT_SEQ-NOT:      DW_AT_META_stmt_sequence
+
+; STMT_SEQ:   [2] DW_TAG_subprogram
```
Comment (pogo59) [inline llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll:9]: I think this `[2]` is the abbrev code? Which is not guaranteed stable; this should use a FileCheck variable.

File: llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+; STMT_SEQ:  	       DW_AT_META_stmt_sequence    DW_FORM_sec_offset
+; STMT_SEQ:   DW_TAG_subprogram [2]
+; STMT_SEQ:       DW_AT_META_stmt_sequence [DW_FORM_sec_offset]	(0x00000028)
+; STMT_SEQ:   DW_AT_name {{.*}}func01
+; STMT_SEQ:   DW_TAG_subprogram [2]
+; STMT_SEQ:       DW_AT_META_stmt_sequence [DW_FORM_sec_offset]	(0x00000038)
+; STMT_SEQ:   DW_AT_name {{.*}}main
```
Comment (pogo59) [inline llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll:16]: The test verifies that the new attribute is emitted and points.... somewhere. It also needs to look at the line table, to show (a) the reference is to the right place, and (b) the extra end_sequence is emitted in the right place.

**Comment by dwblaikie** [comment]
Anyone want to debate whether this should be "DW_AT_META" or "DW_AT_LLVM"? (mostly I think we've used LLVM for things that are usable within the LLVM ecosystem, which it seems the intent is that this might be - llvm-gsym, etc?) I guess the Apple accelerator tables were _APPLE_ but perhaps in part because they were only usable on with the MachO debug info distribution mechanism?

**Comment by dwblaikie** [comment]
Might want/need to start with this patch split up - adding the assembly first, then using it for LLVM code generation.

Which, I think, will mean needing to define some assembly directives to use this functionality in the assembler/assembly code (which, at a very cursory glance of this patch, aren't provided yet)?

While it's true we do emit some DWARF when using the integrated assembler that we cannot emit via assembly (the ability to use multiple separate line tables, I think, in LTO - only done when going direct to machine code via the integrated assembler, and when going via assembly we get one monolithic line table) - it's not an ideal state and this feature's a bit more intrusive than whether or not it's one line table or smaller separate line tables (those are at least roughly semantically equivalent - this new proposed feature isn't just a difference in encoding, but in functionality/content)

**Comment by alx32** [comment]
> adding the assembly first, then using it for LLVM code generation.
> Which, I think, will mean needing to define some assembly directives

Thank you for the input! I am not very familiar with this part of the codebase. Could you provide some pointers to similar existing functionality or tests that would be relevant to what the first part of this change should be?

The most similar feature to the new `DW_AT_META_stmt_sequence` is `DW_AT_stmt_list`. In the assembly tests, `DW_AT_stmt_list` appears represented like this: https://github.com/llvm/llvm-project/blob/main/llvm/test/DebugInfo/X86/abbr_offset.s#L38
```assembly
	.byte 16                      # DW_AT_stmt_list
	...
	.long .Lline_table_start0     # DW_AT_stmt_list
```
It seems like this approach should also work for `DW_AT_META_stmt_sequence`.

Am I missing something, or is there another aspect I should be looking into?

File: llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll (hunk 2 above)
Comment (alx32) [reply llvm/test/DebugInfo/X86/DW_AT_META_stmt_seq_sec_offset.ll:16]: Seems we can rely on the offsets being stable and match them as absolute values: https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/llvm-dwarfdump/X86/verbose.test#L15-L30

**Comment by dwblaikie** [comment]
> > adding the assembly first, then using it for LLVM code generation.
> > Which, I think, will mean needing to define some assembly directives
> 
> Thank you for the input! I am not very familiar with this part of the codebase. Could you provide some pointers to similar existing functionality or tests that would be relevant to what the first part of this change should be?
> 
> The most similar feature to the new `DW_AT_META_stmt_sequence` is `DW_AT_stmt_list`. In the assembly tests, `DW_AT_stmt_list` appears represented like this: [abbr_offset.s:38](https://github.com/llvm/llvm-project/blob/main/llvm/test/DebugInfo/X86/abbr_offset.s#L38)
> 
> ```assembly
> 	.byte 16                      # DW_AT_stmt_list
> 	...
> 	.long .Lline_table_start0     # DW_AT_stmt_list
> ```
> 
> It seems this representation should also work for `DW_AT_META_stmt_sequence` - without additional changes. Am I missing something, or is there another aspect I should be looking into?

So this works for the line table because the stmt_list refers to the very start of the line table - so the assembly code refers to that label, then later on it emits a label:

```
        .section        .debug_line,"",@progbits
.Lline_table_start0:
```
Then the assembler emits the line table during assembly.

So it's only possible to refer to the very start of the line table in this way - to refer into the line table at particular locations, it would require assembler features.
```

**Comment by pogo59** [comment]
> Currently, Clang generates function line information in the debug_line section without directly linking the line entries to their originating functions.

If you compile with -ffunction-sections, then there won't be any sequences in the line table that cross function boundaries. There will always be a DW_LNE_set_address opcode that points back to the start of the function. This pointer is in the opposite direction to what you are looking for, so it requires a bit more analysis to associate sequences with subprograms, but the information is there. Is that sufficient?

**Comment by alx32** [comment]
This feature is primarily designed for MachO, so the following explanation is specific to that context:

> There will always be a DW_LNE_set_address opcode that points back to the start of the function.

You're right that this method works well when we're dealing with a single object file. For instance, in `foo.o`, the line entries for the `foo` function will correctly point to the `foo` instructions based on the address, and the same goes for `bar.o` with the `bar` function.

However, the challenge arises when identical functions like `foo` and `bar` are merged during the linking process with Identical Code Folding (ICF). After linking `foo.o` and `bar.o` into a single binary (`out.bin`) and creating a debug symbol file (`out.dSYM`), there will be only one merged function in `out.bin` representing both `foo` and `bar`. In the `out.dSYM`, the line entries for both `foo` and `bar` will point to this single merged function. At this point, it becomes impossible to determine which line entries originally belonged to `foo` and which to `bar`.

Is there currently a way to attribute the line information in `out.dSYM` to either `foo` or `bar` ?
I think one possibility would be to leave the object files as-is and have `dsymutil` add additional attributes to `out.dSYM` - but I don't think this is an approach we take for anything currently.

**Comment by alx32** [comment]
@dwblaikie
> So it's only possible to refer to the very start of the line table in this way

Oh I see. Would it be possible to specify it as a numeric offset (at least for the purpose of this change) ? i.e:
```
 .long  0x41   # DW_AT_META_stmt_sequence
```

**Comment by pogo59** [comment]
> Anyone want to debate whether this should be "DW_AT_META" or "DW_AT_LLVM"?

That kind of depends on the intended consumer, what it's actually used for, etc. If this is prototyping a way to solve the ICF confusion within debuggers, to be consumed by LLDB, with an eye toward making an actual DWARF proposal, then the LLVM vendor seems appropriate. If this is for some secret tooling within Meta, then META would be most appropriate.

**Comment by pogo59** [comment]
> This patch introduces a new Clang flag, `-mllvm -emit-func-debug-line-table-offsets`

Nit: Technically that is not a Clang flag. A "Clang flag" would be a driver or cc1 option, not `-mllvm`.

**Comment by alx32** [comment]
Here is the RFC with more information: https://discourse.llvm.org/t/rfc-new-dwarf-attribute-for-symbolication-of-merged-functions/79434

I will look into the the assembly attribute generation support.

**Pull request closed without merging**